src/internet/radiobrowser/radiobrowserservice.cpp

```diff
@@ -0,0 +1,390 @@
+const char* RadioBrowserService::kSettingsGroup = "RadioBrowser";
+const char* RadioBrowserService::kSchemeName = "radiobrowser";
+
+QString RadioBrowserService::SearchUrl = "%1/json/stations/byname/%2?limit=%3";
+QString RadioBrowserService::PlayClickUrl = "%1/json/url/%2";
+
+QList<RadioBrowserService::Branch> RadioBrowserService::BranchList = {
+    {"By Country", "%1/json/countries",
+     "%1/json/stations/bycountryexact/%2?hidebroken=true", Type_Category},
+    {"By Language", "%1/json/languages",
+     "%1/json/stations/bylanguageexact/%2?hidebroken=true", Type_Category},
+    {"By Tag", "%1/json/tags", "%1/json/stations/bytagexact/%2?hidebroken=true",
+     Type_Category},
+    {"By Codec", "%1/json/codecs",
+     "%1/json/stations/bycodecexact/%2?hidebroken=true", Type_Category},
+    {"Top 100 Clicked", "",
+     "%1/json/stations/search?order=clickcount&reverse=true&limit=100",
+     Type_Top100},
+    {"Top 100 Voted", "",
+     "%1/json/stations/search?order=votes&reverse=true&limit=100", Type_Top100},
+    {"Top 100 Trending", "",
+     "%1/json/stations/search?order=clicktrend&reverse=true&limit=100",
+     Type_Top100}};
+
+RadioBrowserService::RadioBrowserService(Application* app,
+                                         InternetModel* parent)
+    : InternetService(kServiceName, app, parent, parent),
+      root_(nullptr),
+      context_menu_(nullptr),
+      network_(new NetworkAccessManager(this)),
+      name_(kServiceName),
+      url_scheme_(kSchemeName),
+      url_handler_(new RadioBrowserUrlHandler(app, this, this)),
+      homepage_url_(QUrl("https://www.radio-browser.info")),
+      icon_(IconLoader::Load("radiobrowser", IconLoader::Provider)) {
+  app_->player()->RegisterUrlHandler(url_handler_);
+  app_->global_search()->AddProvider(
+      new RadioBrowserSearchProvider(app_, this, this));
+}
+
+RadioBrowserService::~RadioBrowserService() { delete context_menu_; }
+
+QStandardItem* RadioBrowserService::CreateRootItem() {
+  root_ = new QStandardItem(icon_, name_);
+  root_->setData(true, InternetModel::Role_CanLazyLoad);
+  return root_;
+}
+
+void RadioBrowserService::LazyPopulate(QStandardItem* item) {
+  switch (item->data(InternetModel::Role_Type).toInt()) {
+    case InternetModel::Type_Service:
+      RefreshRootItem();
+      break;
+    case RadioBrowserService::Type_Category:
+      RefreshCategory(item);
+      break;
+    case RadioBrowserService::Type_CategoryItem:
+      RefreshCategoryItem(item);
+      break;
+    case RadioBrowserService::Type_Top100:
+      RefreshTop100(item);
+      break;
+    default:
+      break;
+  }
+}
+
+void RadioBrowserService::RefreshRootItem() {
+  if (!EnsureServerConfig()) {
+    ShowConfig();
+    return;
+  }
+
+  if (root_->hasChildren()) root_->removeRows(0, root_->rowCount());
+  for (auto branch : RadioBrowserService::BranchList) {
+    QStandardItem* item = new QStandardItem(
+        IconLoader::Load("icon_radio", IconLoader::Lastfm), QString());
```

Comment:
Wrong icon?

Reply:
Not really; the radiobrowser icon was too repetitive when having  a lot of items; i saw this generic icon being used by other providers (icecast, digitally, intergalacticfm, ..) and thought about using it. Would you prefer another icon, eg. a folder icon?

Reply:
No, sorry, I was thrown off by the "Lastfm" part. I guess that's how we've been identifying our misc icon folder. Though, maybe this icon really is from lastfm.
https://github.com/clementine-player/Clementine/blob/master/src/ui/iconloader.cpp#L91